WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift

```diff
@@ -0,0 +1,152 @@
+    private let stores: StoresManager
+
+    init(stores: StoresManager = ServiceLocator.stores) {
+        self.stores = stores
+    }
+
+    /// Adds a product and sets its password remotely.
+    /// - Parameters:
+    ///   - product: The product to be added remotely.
+    ///   - password: Optional password of the product.
+    ///   - onCompletion: Called when the remote process finishes.
+    func addProduct(product: EditableProductModel,
+                    password: String?,
+                    onCompletion: @escaping AddProductCompletion) {
+        addProductRemotely(product: product) { productResult in
+            switch productResult {
+            case .failure(let error):
+                // TODO-2766: M4 analytics
+                onCompletion(.failure(error), nil)
+            case .success(let product):
+                // `self` is retained because the use case is not usually strongly held.
+                self.updatePasswordRemotely(product: product, password: password) { passwordResult in
+                    switch passwordResult {
+                    case .failure(let error):
+                        // TODO-2766: M4 analytics
+                        onCompletion(.success(product), .failure(error))
+                    case .success(let password):
+                        // TODO-2766: M4 analytics
+                        onCompletion(.success(product), .success(password))
+                    }
+                }
+            }
+        }
+    }
+
+    /// Edits a product and its password remotely.
+    /// - Parameters:
+    ///   - product: The product to be updated remotely.
+    ///   - originalProduct: The original product before any edits.
+    ///   - password: Optional password of the product.
+    ///   - originalPassword: Optional password of the original product.
+    ///   - onCompletion: Called when the remote process finishes.
+    func editProduct(product: EditableProductModel,
+                     originalProduct: EditableProductModel,
+                     password: String?,
+                     originalPassword: String?,
+                     onCompletion: @escaping EditProductCompletion) {
+        let group = DispatchGroup()
+
+        var productResult: Result<EditableProductModel, ProductUpdateError>?
+        var passwordResult: Result<String?, Error>?
+
+        group.enter()
+        editProductRemotely(product: product, originalProduct: originalProduct) { result in
+            productResult = result
+            group.leave()
+        }
+
+        group.enter()
+        updatePasswordRemotely(product: product, password: password, originalPassword: originalPassword) { result in
+            passwordResult = result
+            group.leave()
+        }
+
+        group.notify(queue: .main) {
+            guard let productResult = productResult, let passwordResult = passwordResult else {
+                assertionFailure("Unexpected nil result after updating product and password remotely")
```

Comment:
my original thought for not passing anything to the completion block was because it's theoretically impossible to reach here (`group.notify` block) without both dispatch blocks complete where both results are set. It might be confusing to read though, I'll update to add a call to `onCompletion`

Reply:
updated in https://github.com/woocommerce/woocommerce-ios/pull/2785/commits/ec900ac19483a08c697256300524b4cf15b86bcf (I created another error case `.unexpected` to differentiate from `.unknown` that is for other errors that have an associated error in https://github.com/woocommerce/woocommerce-ios/pull/2781)